msgraphcore/client_factory.py

```diff
@@ -0,0 +1,64 @@
+import functools
+from typing import Optional
+
+from requests import Session
+
+from msgraphcore.constants import CONNECTION_TIMEOUT, REQUEST_TIMEOUT
+from msgraphcore.enums import APIVersion, NationalClouds
+from msgraphcore.middleware.abc_token_credential import TokenCredential
+from msgraphcore.middleware.authorization import AuthorizationHandler
+from msgraphcore.middleware.middleware import BaseMiddleware, MiddlewarePipeline
+
+
+class HTTPClientFactory:
+    """
+    Constructs HTTP Client(session) instances configured with either custom or default
+    pipeline of middleware.
+    """
+    def __init__(self, **kwargs):
+        """Class constructor that accepts a user provided session object and kwargs
+        to configure the request handling behaviour of the client"""
+        self.api_version = kwargs.get('api_version', APIVersion.v1)
+        self.endpoint = kwargs.get('cloud', NationalClouds.Global)
+        self.timeout = kwargs.get('timeout', (CONNECTION_TIMEOUT, REQUEST_TIMEOUT))
+        self.session = kwargs.get('session', Session())
+
+        self._get_base_url()
+        self._set_default_timeout()
+
+    def create_with_default_middleware(self, credential: TokenCredential, **kwargs) -> Session:
+        """Applies the default middleware chain to the HTTP Client"""
+        middleware = [
+            AuthorizationHandler(credential, **kwargs),
+        ]
+        self._register(middleware)
+        return self.session
+
+    def create_with_custom_middleware(self, middleware: [BaseMiddleware]) -> Session:
+        """Applies a custom middleware chain to the HTTP Client """
+        if not middleware:
+            raise ValueError("Please provide a list of custom middleware")
+        self._register(middleware)
+        return self.session
+
+    def _get_base_url(self):
+        """Helper method to get the base url"""
+        base_url = self.endpoint + '/' + self.api_version
```

Comment:
This method sets base url instead of getting base url. Consider renaming to _set_base_url.

Reply:
Done.